networkx/algorithms/community/louvain.py

```diff
@@ -0,0 +1,531 @@
+"""
+from __future__ import print_function
+__all__ = ["partition_at_level", "modularity", "best_partition", "generate_dendogram", "induced_graph"]
+__author__ = """Thomas Aynaud (thomas.aynaud@lip6.fr)"""
+#    Copyright (C) 2009 by
+#    Thomas Aynaud <thomas.aynaud@lip6.fr>
+#    All rights reserved.
+#    BSD license.
+
+__PASS_MAX = -1
+__MIN = 0.0000001
+
+import networkx as nx
+import sys
+import types
+import array
+
+
+def partition_at_level(dendogram, level) :
+    """Return the partition of the nodes at the given level
+
+    A dendogram is a tree and each level is a partition of the graph nodes.
+    Level 0 is the first partition, which contains the smallest communities, and the best is len(dendogram) - 1.
+    The higher the level is, the bigger are the communities
+
+    Parameters
+    ----------
+    dendogram : list of dict
+       a list of partitions, ie dictionnaries where keys of the i+1 are the values of the i.
+    level : int
+       the level which belongs to [0..len(dendogram)-1]
+
+    Returns
+    -------
+    partition : dictionnary
+       A dictionary where keys are the nodes and the values are the set it belongs to
+
+    Raises
+    ------
+    KeyError
+       If the dendogram is not well formed or the level is too high
+
+    See Also
+    --------
+    best_partition which directly combines partition_at_level and generate_dendogram to obtain the partition of highest modularity
+
+    Examples
+    --------
+    >>> G=nx.erdos_renyi_graph(100, 0.01)
+    >>> dendo = generate_dendogram(G)
+    >>> for level in range(len(dendo) - 1) :
+    >>>     print "partition at level", level, "is", partition_at_level(dendo, level)
+    """
+    partition = dendogram[0].copy()
+    for index in range(1, level + 1) :
+        for node, community in partition.items() :
+            partition[node] = dendogram[index][community]
+    return partition
+
+
+def modularity(partition, graph) :
+    """Compute the modularity of a partition of a graph
+
+    Parameters
+    ----------
+    partition : dict
+       the partition of the nodes, i.e a dictionary where keys are their nodes and values the communities
+    graph : networkx.Graph
+       the networkx graph which is decomposed
+
+    Returns
+    -------
+    modularity : float
+       The modularity
+
+    Raises
+    ------
+    KeyError
+       If the partition is not a partition of all graph nodes
+    ValueError
+        If the graph has no link
+    TypeError
+        If graph is not a networkx.Graph
+
+    References
+    ----------
+    .. 1. Newman, M.E.J. & Girvan, M. Finding and evaluating community structure in networks. Physical Review E 69, 26113(2004).
+
+    Examples
+    --------
+    >>> G=nx.erdos_renyi_graph(100, 0.01)
+    >>> part = best_partition(G)
+    >>> modularity(part, G)
+    """
+    if type(graph) != nx.Graph :
+        raise TypeError("Bad graph type, use only non directed graph")
+
+    inc = dict([])
```

Comment:
Use `inc = defaultdict(lambda: 0)` and the same for `deg`, so that you can just do `inc[com] += weight` below.